README
photogallery - simple static photogallery script

Copyright (C) 2007,2008,2009  Christian Garbs <mitch@cgarbs.de>
licensed under GNU GPL v2 or later (see COPYING)

Homepage:
http://www.cgarbs.de/photogallery.en.html


For the main instructions see the top of the photogallery script.


The other scripts are addons:

== editcomments
Conveniently edit picture comments.  Give the pictures as parameters.

== exif2text
Include standard EXIF tags in picture comments.  Give the pictures as
parameters.

== photogallery_diff.pl
List changes within photogallery directory structure.
Configuration is read from ~/.photogallery-conf.pl, see
photogallery-conf.pl.sample for an example.

== photogallery_diff2html
Convert change list generated by photogallery_diff into an HTML page.

== photogallery_diff2rss
Convert change list generated by photogallery_diff into an RSS feed.



License:



HISTORY
photogallery HISTORY


photogallery v1.30 - 2009-04-13
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* bugfix: broken url encoding with bash 3.2
* clarify copyright


photogallery v1.29 - 2009-04-12
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* bugfix: don't print help text when no arguments are given


photogallery v1.28 - 2008-12-27
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* bugfix: urlencode special characters
* bugfix: broken HTML without CSS
* bugfix: handle animated GIFs properly
* bugfix: speed up exif2text
* support video files
* include ISO in exif2text
* provide help texts
* version number handling


photogallery v1.27 - 2008-07-16
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* fix inclusion of version number in gallery footer
* add 'exif2text' to include EXIF tags in comments
* rename scripts to allow proper Debian packaging


photogallery v1.26 - 2008-07-13
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* create proper release tarball including all tools and short README


photogallery v1.25 - 2008-07-08
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* begin the HISTORY file


editcomments.sh
#!/bin/bash
#
# 2008 (c) by Christian Garbs <mitch@cgarbs.de>
#
# edit comments for photogallery.sh gallery
# licensed under GNU GPL v2
#

TEMP1=$(mktemp)
TEMP2=$(mktemp)

# usage note
if [ "${*}" = "" ] ; then
    echo "usage: editcomments.sh FILE..."
    exit 1
fi

# collect files and comments
for FILE in "${@}"; do
    if [ -e "$FILE" ] ; then
	echo "====$FILE"
	[ -e "$FILE.text" ] && cat "$FILE.text"
    fi
done > $TEMP1

# edit list
if [ ! -s $TEMP1 ] ; then
    echo "no files found"
    exit 1
fi
$EDITOR $TEMP1

# add end mark
echo "====" >> $TEMP1

# split list into files
touch $TEMP2
while read LINE ; do

        if [ "${LINE:0:4}" = '====' ] ; then
	    rm -f "$FILENAME"
	    [ -s $TEMP2 ] && mv $TEMP2 "$FILENAME"
	    touch $TEMP2
		
	    FILENAME="${LINE:4}"
	    if [ ! -e "${FILENAME}" -a "$FILENAME" ] ; then
		echo "skipping file $FILENAME: does not exist"
	    fi
	    FILENAME="${FILENAME}.text"
	else
	    echo $LINE >> $TEMP2
	fi

done < $TEMP1

rm $TEMP1 $TEMP2


photogallery.sh
#!/bin/bash
#
# simple static photogallery script
# 2007-2008 (c) by Christian Garbs <mitch@cgarbs.de>
# licensed under GNU GPL v2

#
# usage:
#    photogallery.sh [gallerytitle]
#
# Run this script inside a directory with pictures.
# It will generate a subfolder (set via $SUBDIR) with some thumbnails
# and an HTML index file (set via $INDEX).
# The gallery title is optional and defaults to the current directory
# name.
# To describe a gallery, put a file named README into the directory.
# It will be included in the generated HTML index file.
# To describe an image some.jpeg, put a file named some.jpeg.text into
# the same directory.
# 
# Already existing thumbnails will not be overwritten.  If you change
# the thumbnail sizes ($MEDIUM and $SMALL) you need to remove the
# thumbnail subfolder before you can run photogallery.sh again.
#
# Basic CSS support is available:
# The environment variables $GALLERYCSS and $THUMBCSS are evaluated
# and directly included in the generated HTML files.  Tag IDs are
# #nav, #pic, #sig and #txt.  In a gallery, you can also set img.txt
# and img.notxt to distinguish pictures with comments from those
# tihous.
#
# Example:
# GALLERYCSS="#txt,#sig{float:right;}#pic{clear:left;}#nav{float:left;}"
#
# To generate lots of galleries in subdirectories use this (only
# works if there are no line breaks in your directory names):
# 
#   find . -depth \! -name .\* -type d | \
#   while read DIR; do ( echo "$DIR"; cd "$DIR"; photogallery.sh ) ; done
#
# If you want to browse your galleries directly from your filesystem
# without a webserver, you have to run this twice so that all links
# are generated properly!
#
#
# External commands needed:
#    convert from Imagemagick
#    dcraw (only if you want to convert RAW files)
#    sed
#

#### commandline parameters

TITLE=${1:-${PWD##*/}}

#### configuration

CHARSET=UTF-8
DATELANG=C
MEDIUM=640x640
SMALL=20%
SUBDIR=.webthumbs
INDEX=index.html

VERSION=1.26git

#### function declarations

status() {
    echo "$@" 1>&2
}

html_head() {
    CSS="$1"
    echo '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml1.dtd">'
    echo '<html><head>'
    echo "<title>$TITLE</title>"
    if [ "$CSS" ] ; then
	echo '<style type="text/css"><!--'
	echo $CSS
	echo '--></style>'
    fi
    echo "<meta http-equiv=\"content-type\" content=\"text/html; charset=${CHARSET}\" />"
    echo '</head>'
}

html_foot() {
    echo '<p id="sig"><small><small><i>generated on ' "$(LANG=${DATELANG} date)" 'by <a href="http://www.cgarbs.de/photogallery.en.html">photogallery.sh '"v${VERSION}</a></i></small></small></p></body></html>"
}

#### main script

exec > $INDEX

html_head "$GALLERYCSS"
echo '<body>'

if [ -r "../$INDEX" ] ; then
    echo -n "<p id=\"nav\"><a href=\"../$INDEX\">[ up ]</a>"
else
    echo -n '<p id="nav"><a href="..">[ up ]</a>'
fi

SUBDIRS=0
for DIR in *; do

    [ -d "$DIR" ] || continue
    [ -r "$DIR" ] || continue
    [ "${DIR:0:1}" = '.' ] && continue

    echo '<br />'
    if [ -r "$DIR/$INDEX" ] ; then
	echo -n "<a href=\"$DIR/$INDEX\">$DIR/</a>"
    else
	echo -n "<a href=\"$DIR\">$DIR/</a>"
    fi

    status -n :
    let SUBDIRS++

done
echo '</p>'

status -n "$SUBDIRS "

if [ -e README ] ; then
    if [ "$GALLERYCSS" ] ; then
	echo '<div id="txt">'
	cat < README
	echo '</div>'
    else
	echo '<p id="txt">'
	cat < README
	echo '</p>'
    fi
fi

mkdir -p "$SUBDIR" || exit 1

echo '<p id="pic">'

declare -a FILES
for FILE in *; do
    [ -f "$FILE" ] || continue
    [ -r "$FILE" ] || continue
    EXT="${FILE/*.}"

    case "$EXT" in 
	gif|jpeg|jpg|JPG|png|tif|tiff)
	    ;;
	pef|PEF)
	    ;;
	*)
	    continue
	    ;;
    esac

    FILES+=("$FILE")
done

status -n "${#FILES[*]}>"

for (( IDX=0; $IDX < ${#FILES[*]}; IDX+=1 )) ; do

    FILE="${FILES[${IDX}]}"

    EXT="${FILE/*.}"
    M_INDEX="${FILE}_m.html"
    M_FILE="${FILE}_m.jpg"
    S_FILE="${FILE}_s.jpg"

    if [ ! -s "$SUBDIR/$M_FILE" ] ; then
	case "$EXT" in
	    
	    gif|jpeg|jpg|JPG|png|tif|tiff)
		convert -resize $MEDIUM -sharpen 1x0.5 "$FILE" "$SUBDIR/$M_FILE"
		;;
	    
	    pef|PEF)
		FLIP=0
		if [ -e "$FILE.rotation" ] ; then
		    case $(< "$FILE.rotation") in
			180)  FLIP=3 ;;
			90)  FLIP=6 ;;
			270)  FLIP=5 ;;
		    esac
		fi
		dcraw -c -w -o1 -h -t $FLIP "$FILE" | convert -resize $MEDIUM -sharpen 1x0.5 ppm:- "$SUBDIR/$M_FILE"
		;;
	    
	    *)
		continue
		;;
	    
	esac
    fi

    [ -s "$SUBDIR/$S_FILE" ] || convert -resize $SMALL "$SUBDIR/$M_FILE" -sharpen 1x0.5 "$SUBDIR/$S_FILE"

    chmod --reference="$FILE" "$SUBDIR/$M_FILE"
    chmod --reference="$FILE" "$SUBDIR/$S_FILE"

    if [ -s "${FILE}.text" ] ; then
	FILETEXT="$(<"${FILE}.text")"
	ALTTEXT="$FILETEXT"
    else
	FILETEXT=
	ALTTEXT="${FILE}"
    fi
    ALTTEXT=$( echo "$ALTTEXT" | sed -e 's/</\&lt;/g' -e 's/>/\&gt;/g' -e 's/"/\&quot;/g' )

    if [ "$GALLERYCSS" ] ; then
	if [ "$FILETEXT" ] ; then
	    INSERT='class="txt"'
	else
	    INSERT='class="notxt"'
	fi
    else
	if [ "$FILETEXT" ] ; then
	    INSERT=3
	else
	    INSERT=1
	fi
    fi
    echo "<a href=\"$SUBDIR/$M_INDEX\"><img $INSERT alt=\"\" title=\"$ALTTEXT\" src=\"$SUBDIR/$S_FILE\" /></a>"

    (
	html_head "$THUMBCSS"
	echo '<body onLoad="document.getElementById('"'next'"').focus();">'
	echo -n "<p id=\"nav\">"
	PREV=$(( $IDX - 1 ))
	NEXT=$(( $IDX + 1 ))
	if [ $IDX -gt 0 ] ; then
	    echo -n "<p><a href=\"${FILES[${PREV}]}_m.html\">&lt;&lt;</a>"
	else
	    echo -n "<p>&lt;&lt;"
	fi
	echo " &nbsp; <a href=\"../$INDEX\">back</a> &nbsp; "
	if [ $NEXT -lt ${#FILES[*]} ] ; then
	    echo "<a id=\"next\" href=\"${FILES[${NEXT}]}_m.html\">&gt;&gt;</a></p>"
	else
	    echo "&gt;&gt;</p>"
	fi
	echo "</p>"
	echo "<p id=\"pic\"><a href=\"../$FILE\"><img alt=\"$FILE\" src=\"$M_FILE\" /></a></p>"

	if [ "$FILETEXT" ] ; then
	    if [ "$THUMBCSS" ] ; then
		echo "<div id=\"txt\">${FILETEXT}</div>"
	    else
		echo "<p id=\"txt\">${FILETEXT}</p>"
	    fi
	fi

	html_foot
	
    ) > "$SUBDIR/$M_INDEX"

    if [ $(( $IDX % 10 )) == 9 ]; then
	status -n $(( $IDX + 1 ))
    else
	status -n .
    fi

done

echo '</p>'
echo '<!--'
echo PICTURES=${#FILES[*]}
echo SUBDIRS=$SUBDIRS
echo DATE=`date +%Y%m%d%H%M%S`
echo '-->'

html_foot

if [ ${#FILES[*]} == 0 ] ; then
    rmdir "$SUBDIR"
fi

status


photogallery_diff.pl
#!/usr/bin/perl
#
# list changes within photogallery.sh directory structure
#
# 2007 (c) by Christian Garbs <mitch@cgarbs.de>
# licensed under GNU GPL v2
#

use strict;
use warnings;

# parse configuration file
my %conf;
eval(`cat ~/.photogallery-conf.pl`);

# check variables
foreach (qw( DATADIR GALLERYDIR MAXCHANGES )) {
    die "configuration key $_ is missing\n" unless exists $conf{$_};
}

# scan structure
# get new counts
my %count_new;

open FIND, "find $conf{GALLERYDIR} -type f -name index.html|" or die "can't spawn find: $!";
while (my $index = <FIND>) {
    chomp $index;

    next if $index =~ /$conf{SKIPDIR}/;

    open INDEX, '<', $index or die "can't read `$index': $!";
    my $new = { 'SUBDIRS' => 0 };
    while (my $line = <INDEX>) {
	if ($line =~ /^PICTURES=(\d+)$/) {
	    $new->{'PICTURES'} = $1;
	} elsif ($line =~ /^SUBDIRS=(\d+)$/) {
	    $new->{'SUBDIRS'} = $1;
	} elsif ($line =~ /^DATE=(\d+)$/) {
	    $new->{'DATE'} = $1;
	}
    }
    close INDEX or die "can't close `$index': $!";
    $count_new{$index} = $new if exists $new->{'PICTURES'};
}
close FIND or die "can't close find: $!";

# get old counts
my %count_old;
my $oldstats = "$conf{DATADIR}/last_run";

open OLDSTATS, '<', $oldstats or die "can't open `$oldstats': $!";
while (my $line = <OLDSTATS>) {
    chomp $line;
    if ($line =~ /^(\d+)\t(.+)$/) {
	$count_old{$2}->{'PICTURES'} = $1;
    }
}
close OLDSTATS or die "can't close `$oldstats': $!";

# get old diffs
my @diff;
my $olddiffs = "$conf{DATADIR}/last_diffs";

open OLDDIFFS, '<', $olddiffs or die "can't open `$olddiffs': $!";
while (my $line = <OLDDIFFS>) {
    chomp $line;
    if ($line =~ /^(\d+)\t(\d+)\t(\d+)\t(.+)$/) {
	push @diff, $line;
    }
}
close OLDDIFFS or die "can't close `$olddiffs': $!";


# generate diffs
my %count_change;
#   add
foreach my $dir (sort keys %count_new) {
    my ($old, $new) = ($count_old{$dir}, $count_new{$dir});
    if (defined $old) {
	if ($old->{'PICTURES'} != $new->{'PICTURES'}) {
	    $count_change{$dir} = { 'OLD' => $old, 'NEW' => $new };
	}
    } else {
	$count_change{$dir} = { 'OLD' => { 'PICTURES' => 0 }, 'NEW' => $new };
    }
}
#   remove
foreach my $dir (sort keys %count_old) {
    my ($old, $new) = ($count_old{$dir}, $count_new{$dir});
    if (defined $new) {
	if ($old->{'PICTURES'} != $new->{'PICTURES'}) {
	    $count_change{$dir} = { 'OLD' => $old, 'NEW' => $new };
	}
    } else {
	$count_change{$dir} = { 'OLD' => $old, 'NEW' => { 'PICTURES' => 0 } };
    }
}

# print diffs
my $timestamp = `date +%Y%m%d%H%M%S`;
chomp $timestamp;
foreach my $dir (reverse sort keys %count_change) {
    push @diff, sprintf "%d\t%d\t%s\t%s",
    $count_change{$dir}->{'OLD'}->{'PICTURES'},
    $count_change{$dir}->{'NEW'}->{'PICTURES'},
    $timestamp,
#    exists $count_change{$dir}->{'NEW'}->{'DATE'} ? $count_change{$dir}->{'NEW'}->{'DATE'} : $timestamp,
    $dir;
}

# chop diffs
if (scalar @diff > $conf{MAXCHANGES}) {
    splice(@diff, 0, -$conf{MAXCHANGES});
}

### TODO below here

# print new diffs
my $newdiffs = "$conf{DATADIR}/last_diffs.new";
open NEWDIFFS, '>', $newdiffs or die "can't open `$newdiffs': $!";
foreach my $diff (@diff) {
    print NEWDIFFS "$diff\n";
}
close NEWDIFFS or die "can't close `$newdiffs': $!";

# print new status
my $newstats = "$conf{DATADIR}/last_run.new";
open NEWSTATS, '>', $newstats or die "can't open `$newstats': $!";
foreach my $dir (sort keys %count_new) {
    printf NEWSTATS "%d\t%s\n",
    $count_new{$dir}->{'PICTURES'},
    $dir;
}
close NEWSTATS or die "can't close `$newstats': $!";

# rename
my $bakdiffs = "$conf{DATADIR}/last_diffs.bak";
my $bakstats = "$conf{DATADIR}/last_run.bak";

my @moves = (
	     $olddiffs, $bakdiffs,
	     $oldstats, $bakstats,
	     $newdiffs, $olddiffs,
	     $newstats, $oldstats
	     );

while (@moves) {
    my ($from, $to) = splice(@moves, 0, 2);
    rename $from, $to or die "can't rename `$from' to `$to': $!";
}


photogallery_diff2html.pl
#!/usr/bin/perl
#
# converts photogallery.sh changes to HTML page
#
# 2007 (c) by Christian Garbs <mitch@cgarbs.de>
# licensed under GNU GPL v2
#

use strict;
use warnings;

# parse configuration file
my %conf;
eval(`cat ~/.photogallery-conf.pl`);

# check variables
foreach (qw( DATADIR DIRPREFIX WEBPREFIX RSSURL CHARSET DATELANG )) {
    die "configuration key $_ is missing\n" unless exists $conf{$_};
}

# get old diffs
my @diff;
my $diffs = "$conf{DATADIR}/last_diffs";

open DIFFS, '<', $diffs or die "can't open `$diffs': $!";
while (my $line = <DIFFS>) {
    chomp $line;
    if ($line =~ /^(\d+)\t(\d+)\t(\d+)\t(.+)$/) {
	push @diff, {
	    'OLD' => $1,
	    'NEW' => $2,
	    'DATE' => $3,
	    'PATH' => $4
	};
    }
}
close DIFFS or die "can't close `$diffs': $!";

# print header
print <<"EOF";
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml1.dtd">
<html><head>
<title>photogallery updates</title>
<meta http-equiv="content-type" content="text/html; charset=${conf{CHARSET}}" />
<link rel="alternate" type="application/rss+xml" title="RSS-Feed" href="$conf{RSSURL}">
</head><body>
<h1>photogallery updates</h1>
EOF
    ;

# print changes
print "<ul>\n";
foreach my $diff (reverse @diff) {
    my $date = $diff->{DATE};
    $date =~ s/^(\d\d\d\d)(\d\d)(\d\d)(\d\d)(\d\d).*$/$3.$2.$1 $4:$5h/;
    my $path = $diff->{PATH};
    $path =~ s/^$conf{DIRPREFIX}//;
    my $url = $conf{WEBPREFIX}.$path;
    $path =~ s:/index.html$::;

    my $text;

    if ($diff->{'OLD'} == 0) {
	$text = "neuer Ordner mit $diff->{'NEW'} Bildern";
    } elsif ($diff->{'NEW'} == 0) {
	# gelöscht, so what?
    } elsif ($diff->{'NEW'} > $diff->{'OLD'}) {
	$text = ($diff->{'NEW'} - $diff->{'OLD'}) . " neue Bilder";
    } else {
	$text = ($diff->{'OLD'} - $diff->{'NEW'}) . " Bilder gelöscht";
    }

    printf "<li>%s <a href=\"%s\">%s</a><br />%s<br />&nbsp;</li>\n",
    $date,
    $url,
    $path,
    $text
	if defined $text;
}
print "</ul>\n";

# print footer
my $date = `LANG=${conf{DATELANG}} date`;
chomp $date;
print "<p>(<a href=\"$conf{RSSURL}\">RSS feed</a>)</p>\n";
print "<p><small><small><i>generated on $date by ";
print '$Id: photogallery_diff2html.pl,v 1.7 2007-08-12 17:29:21 mitch Exp $';
print "</i></small></small></p></body></html>\n";


photogallery_diff2rss.pl
#!/usr/bin/perl
#
# converts photogallery.sh changes to HTML page
#
# 2007 (c) by Christian Garbs <mitch@cgarbs.de>
#

use strict;
use warnings;

use Digest::MD5 qw(md5_hex);
use POSIX qw(strftime);

# parse configuration file
my %conf;
eval(`cat ~/.photogallery-conf.pl`);

# check variables
foreach (qw( DATADIR DIRPREFIX WEBPREFIX CHARSET CHANGESURL RSSTITLE RSSAUTHOR )) {
    die "configuration key $_ is missing\n" unless exists $conf{$_};
}

# get old diffs
my @diff;
my $diffs = "$conf{DATADIR}/last_diffs";

open DIFFS, '<', $diffs or die "can't open `$diffs': $!";
while (my $line = <DIFFS>) {
    chomp $line;
    if ($line =~ /^(\d+)\t(\d+)\t(\d+)\t(.+)$/) {
	push @diff, {
	    'OLD' => $1,
	    'NEW' => $2,
	    'DATE' => $3,
	    'PATH' => $4
	};
    }
}
close DIFFS or die "can't close `$diffs': $!";

# print header
print <<"EOF";
<?xml version="1.0" encoding="${conf{CHARSET}}"?>
<rss version="2.0"
 xmlns:dc="http://purl.org/dc/elements/1.1/"
 xmlns:content="http://purl.org/rss/1.0/modules/content/">
  <channel>
    <title>$conf{RSSTITLE}</title>
    <link>$conf{CHANGESURL}</link>
    <description>$conf{RSSTITLE}</description>
EOF
;
print '    <generator>$Id: photogallery_diff2rss.pl,v 1.4 2007-08-26 08:06:13 mitch Exp $</generator>';
print "\n";

# print changes
my $timezone = `date +%z`;
chomp $timezone;
foreach my $diff (reverse @diff) {
    my $date = $diff->{DATE};
    $date = strftime("%a, %d %b %Y %H:%M:%S $timezone", substr($date,12,2), substr($date,10,2), substr($date,8,2), substr($date,6,2), substr($date,4,2)-1, substr($date,0,4)-1900);
    my $path = $diff->{PATH};
    $path =~ s/^$conf{DIRPREFIX}//;
    my $url = $conf{WEBPREFIX}.$path;
    $path =~ s:/index.html$::;

    my $text;

    if ($diff->{'OLD'} == 0) {
	$text = "neuer Ordner mit $diff->{'NEW'} Bildern";
    } elsif ($diff->{'NEW'} == 0) {
	# gelöscht, so what?
    } elsif ($diff->{'NEW'} > $diff->{'OLD'}) {
	$text = ($diff->{'NEW'} - $diff->{'OLD'}) . " neue Bilder";
    } else {
	$text = ($diff->{'OLD'} - $diff->{'NEW'}) . " Bilder gelöscht";
    }

    next unless defined $text;

    # get thumbnail
    my $pic;
    my $thumbdir = $diff->{PATH};
    $thumbdir =~ s:index.html:.webthumbs/:;
    if ( -d $thumbdir ) {
	opendir THUMBS, $thumbdir or die "can't opendir `${thumbdir}': $!";
	while (my $file = readdir (THUMBS)) {
	    if ($file =~ /_s.jpg$/) {
		$pic = $file;
		last;
	    }
	}
	closedir THUMBS or die "can't closedir `${thumbdir}': $!";
	if ($pic) {
	    $thumbdir =~ s/^$conf{DIRPREFIX}/$conf{WEBPREFIX}/;
	    $text .= "<p><img src=\"$thumbdir$pic\"></p>";
	}
    }

    my $guid = md5_hex( $url . $path . $text . $date );

    printf <<"EOF"
    <item>
      <title><![CDATA[${path}]]></title>
      <content:encoded>\n<![CDATA[${text}]]></content:encoded>
      <pubDate>$date</pubDate>
      <dc:creator>$conf{RSSAUTHOR}</dc:creator>
      <guid isPermaLink=\"false\">${guid}</guid>
      <link>${url}</link>
    </item>
EOF
;
}

# print footer
print "  </channel>\n";
print "</rss>\n";
